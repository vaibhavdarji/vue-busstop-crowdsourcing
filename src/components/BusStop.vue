<template>
  <v-hover>
    <v-card  tile :class="['rounded-card', `elevation-${hover ? 12 : 2}`]" slot-scope="{ hover }">
      <v-card-title primary-title>
        <v-icon left>edit_location</v-icon>
        <div class="headline">{{stop.name}}</div>
      </v-card-title>
      <v-divider></v-divider>
      <v-card-actions>
        <v-flex xs8>
          <v-list-tile-title>
            <span class="headline">$ {{stop.donationsRaisedInDollars}}</span>
            <span class="body-1 pl-2">{{Constants.CURRENCY}} raised</span>
          </v-list-tile-title>
          <v-list-tile-sub-title>
            <v-progress-linear
              :value="percentage"
              :background-color="Constants.LINEAR_PROGRESS_BAR.BG_COLOR"
              :color="Constants.LINEAR_PROGRESS_BAR.COLOR"
            ></v-progress-linear>
          </v-list-tile-sub-title>
          <v-list-tile-title>
            <span class="subheading">{{`${percentage} %`}}</span>
            <span class="pl-2 body-1">of {{`${Constants.CURRENCY} ${Constants.MAX_DONATE}`}} raised</span>
            <v-tooltip top :color="Constants.LINEAR_PROGRESS_BAR.COLOR" max-width="210px">
              <template v-slot:activator="{ on }">
                <span class="ml-1 body-1 text-underline" v-on="on">fixed goal</span>
              </template>
              <span>
                This campaign 
                {{stop.donationsRaisedInDollars > 700 ? 'has': 'has not'}}
                reached its goal
                and will receive all funds raised.
              </span>
            </v-tooltip>
          </v-list-tile-title>
        </v-flex>
        <v-flex xs4>
          <custom-button @onClick="selectStop(stop)">Donate</custom-button>
        </v-flex>
      </v-card-actions>
    </v-card>
  </v-hover>
</template>

<script>
import { mapGetters, mapActions } from "vuex";
import Constants from "@/constants";
import CustomButton from "@/components/Button.vue";
import {
  SET_SELECTED_STOP
} from '@/actions/types';
export default {
  name: "BusStop",
  props: {
    stop: {
      type: Object,
      default: {}
    }
  },

  components: {
    CustomButton
  },

  computed: {
    Constants() {
      return Constants;
    },

    percentage() {
      return Math.round(
        (this.stop.donationsRaisedInDollars / Constants.MAX_DONATE) * 100
      );
    }
  },

  methods: {
    ...mapActions({
      "setSelectedStop": SET_SELECTED_STOP
    }),
    selectStop(stop) {
      this.setSelectedStop(stop);
      this.$router.push({ path: "stop" });
    }
  }
};
</script>

<style>
.text-underline {
  border-bottom: 1px dashed;
}
.rounded-card {
  border-radius: 10px;
}
</style>
