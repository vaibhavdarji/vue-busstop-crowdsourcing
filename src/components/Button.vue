<template>
  <v-btn 
    :type="btnType"
    :loading="isLoading"
    v-bind="{ [`${Constants.BUTTON.SIZE}`]: true }"
    :color="Constants.BUTTON.BG_COLOR"
    :class="[`${Constants.BUTTON.COLOR}--text`]"
    :disabled="disabled"
    @click="onClickBtn"
  >
    <slot></slot>
  </v-btn>
</template>

<script>
import Constants from '@/constants';
export default {
  name:'CustomButton',
  props: {
    isLoading: {
      type: Boolean,
      default: false
    },
    btnType: {
      type: String,
      default: 'button' 
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {};
  },
  computed: {
    Constants () {
      return Constants;
    }
  },
  methods: {
    onClickBtn () {
      this.$emit('onClick');
    } 
  }
}
</script>

