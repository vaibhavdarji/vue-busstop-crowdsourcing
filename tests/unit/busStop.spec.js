import { shallowMount, mount } from "@vue/test-utils";
import BusStop from '@/components/BusStop.vue';
import Button from '@/components/Button.vue';
import Vue from "vue";
import Vuex from "vuex";
import Vuetify from "vuetify";
import {
  SET_SELECTED_STOP
} from '@/actions/types';

Vue.use(Vuex);
Vue.use(Vuetify);

const dummyData = [
  { stopId: 1, lat: 33.760262, lng: -84.384706, donationsRaisedInDollars: 0, name: 'Hertz at Portman Blvd' },
  { stopId: 2, lat: 33.760138, lng: -84.388043, donationsRaisedInDollars: 0, name: 'Peachtree Center Mall' }
];

describe("BusStops.vue", () => {
  let actions, store;
  beforeEach(() => {
    actions = {
    };

    actions[SET_SELECTED_STOP] = jest.fn();
    store = new Vuex.Store({
      actions
    });
  });

  it("should match snapshot", () => {
    const wrapper = shallowMount(BusStop, {
      propsData: {
        stop: dummyData[Math.round(Math.random()*(dummyData.length - 1))]
      },
      store,
      Vue
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should contain Button.vue as a child component", () => {
    document.body.setAttribute('data-app', true);
    const wrapper = mount(BusStop, {
      propsData: {
        stop: dummyData[Math.round(Math.random()*(dummyData.length - 1))]
      },
      store,
      Vue
    });
    expect(wrapper.contains(Button)).toBe(true);
  });

  it("should validate selected bus stop data in dom elements", () => {
    const currentStop = dummyData[Math.round(Math.random()*(dummyData.length - 1))];
    const wrapper = mount(BusStop, {
      propsData: {
        stop: currentStop
      },
      store,
      Vue
    });

    // stop title name validation
    expect(wrapper.find('div.headline').text()).toEqual(currentStop.name);

    // stop donation amount validation
    expect(wrapper.find('span.headline').text()).toContain(currentStop.donationsRaisedInDollars);

    // stop donation percentage validation
    expect(wrapper.find('span.subheading').text()).toContain(wrapper.vm.percentage);
  });

  it("should trigger select stop buttion click event. calls store action", () => {
    const currentStop = dummyData[Math.round(Math.random()*(dummyData.length - 1))];
    const wrapper = mount(BusStop, {
      propsData: {
        stop: currentStop
      },
      store,
      Vue
    });
    
    wrapper.setMethods({"selectStop": actions[SET_SELECTED_STOP]});
    const btnInput = wrapper.find('button');
    btnInput.trigger('click');
    expect(actions[SET_SELECTED_STOP]).toHaveBeenCalled();
  });
});