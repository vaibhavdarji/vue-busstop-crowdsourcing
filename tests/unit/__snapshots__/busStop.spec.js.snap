// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusStops.vue should match snapshot 1`] = `
<v-hover-stub
  closedelay="0"
  opendelay="0"
/>
`;
